quark/mdk-2.0.q

```diff
@@ -292,38 +313,116 @@ namespace mdk {
+            }
+
+            targets.add({"version": version, "target": target, "targetVersion": targetVersion});
+        }
+
+        void trace(String level) {
+            set("trace", level);
+        }
+
+        static int _level(String level) {
+            if (_levels.contains(level)) {
+                return _levels[level];
+            } else {
+                return 0;
+            }
+        }
+
+        bool _enabled(String level) {
+            int ilevel = _level("INFO");
+            if (has("trace")) {
+                ilevel = _level(?get("trace"));
+            }
+
+            return _level(level) <= ilevel;
+        }
+
         void _log(String level, String category, String text) {
             _mdk._tracer.setContext(_context);
             _mdk._tracer.log(_mdk.procUUID, level, category, text);
         }
 
         void critical(String category, String text) {
             // XXX: no critical
-            _mdk.logger.error(category + ": " + text);
-            _log("CRITICAL", category, text);
+            if (_enabled("CRITICAL")) {
+                _mdk.logger.error(category + ": " + text);
+                _log("CRITICAL", category, text);
+            }
         }
 
         void error(String category, String text) {
-            _mdk.logger.error(category + ": " + text);
-            _log("ERROR", category, text);
+            if (_enabled("ERROR")) {
+                _mdk.logger.error(category + ": " + text);
+                _log("ERROR", category, text);
+            }
         }
 
         void warn(String category, String text) {
-            _mdk.logger.warn(category + ": " + text);
-            _log("WARN", category, text);
+            if (_enabled("WARN")) {
+                _mdk.logger.warn(category + ": " + text);
+                _log("WARN", category, text);
+            }
         }
 
         void info(String category, String text) {
-            _mdk.logger.info(category + ": " + text);
-            _log("INFO", category, text);
+            if (_enabled("INFO")) {
+                _mdk.logger.info(category + ": " + text);
+                _log("INFO", category, text);
+            }
         }
 
         void debug(String category, String text) {
-            _mdk.logger.debug(category + ": " + text);
-            _log("DEBUG", category, text);
+            if (_enabled("DEBUG")) {
+                _mdk.logger.debug(category + ": " + text);
+                _log("DEBUG", category, text);
+            }
         }
 
         Promise _resolve(String service, String version) {
+            Map<String,List<Map<String,String>>> routes = ?get("routes");
+            if (routes != null && routes.contains(service)) {
+                List<Map<String,String>> targets = routes[service];
+                int idx = 0;
+                while (idx < targets.size()) {
+                    Map<String,String> target = targets[idx];
+                    if (versionMatch(target["version"], version)) {
```

Comment:
The point of doing the version match here is that without it we will end up rerouting requests for all versions of a service (including requests for different major numbers) to the same place, and this would result in a pathological topology since no single service could implement both major versions at the same time.